raft/src/main/java/org/apache/kafka/raft/ReplicatedCounter.java

```diff
@@ -27,110 +28,94 @@
 import org.slf4j.Logger;
 
 import java.nio.ByteBuffer;
+import java.util.OptionalInt;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.atomic.AtomicInteger;
 
-/**
- * This is the simplest interesting state machine. It maintains a simple counter which can only
- * be incremented by one.
- */
-public class ReplicatedCounter implements ReplicatedStateMachine {
+public class ReplicatedCounter {
+    private final int localBrokerId;
     private final Logger log;
-    private final int nodeId;
+    private final RaftClient client;
+
     private final AtomicInteger committed = new AtomicInteger(0);
-    private final boolean verbose;
+    private final AtomicInteger uncommitted = new AtomicInteger(0);
     private OffsetAndEpoch position = new OffsetAndEpoch(0, 0);
-    private AtomicInteger uncommitted;
-    protected RecordAppender appender = null;
+    private LeaderAndEpoch currentLeaderAndEpoch = new LeaderAndEpoch(OptionalInt.empty(), 0);
 
-    public ReplicatedCounter(int nodeId,
-                             LogContext logContext,
-                             boolean verbose) {
-        this.nodeId = nodeId;
+    public ReplicatedCounter(int localBrokerId,
+                             RaftClient client,
+                             LogContext logContext) {
+        this.localBrokerId = localBrokerId;
+        this.client = client;
         this.log = logContext.logger(ReplicatedCounter.class);
-        this.verbose = verbose;
     }
 
-    @Override
-    public void initialize(RecordAppender recordAppender) {
-        appender = recordAppender;
+    private Records tryRead(long durationMs) {
+        CompletableFuture<Records> future = client.read(position, Isolation.COMMITTED, durationMs);
+        try {
+            return future.get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
     }
 
-    @Override
-    public synchronized void becomeLeader(int epoch) {
-        uncommitted = new AtomicInteger(committed.get());
-    }
+    private void apply(Record record) {
```

Comment:
In the approved KIP we still have a field called `AppliedOffset` in the state file to remember the progress of the state machine data apply. Is there a way we could hook it up to the RaftClient?

Reply:
In fact, we probably could defer it to the snapshot implementation, so that the Raft implementation is fully isolated.

Reply:
Yeah, good point. We probably want to take this out of the KIP and leave it to the state machine to manage this.